WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,238 @@
+		DownloadNewVersion = downloadNewVersion;
+	}
+
+	public async void UpdateStatusChangedAsync(UpdateStatus updateStatus)
+	{
+		var tries = 0;
+		bool updateAvailable = !updateStatus.ClientUpToDate || !updateStatus.BackendCompatible;
+		Version targetVersion = updateStatus.ClientVersion;
+		if (!updateAvailable)
+		{
+			return;
+		}
+		if (DownloadNewVersion)
+		{
+			Logger.LogInfo($"Trying to download new version: {targetVersion}");
+			do
+			{
+				tries++;
+				try
+				{
+					bool isReadyToInstall = await GetInstallerAsync(targetVersion).ConfigureAwait(false);
+					Logger.LogInfo($"Version {targetVersion} downloaded successfuly.");
+					updateStatus.IsReadyToInstall = isReadyToInstall;
+					break;
+				}
+				catch (Exception ex)
+				{
+					Logger.LogError($"Geting version {targetVersion} failed with error.", ex);
+				}
+			} while (tries < MaxTries);
+		}
+
+		UpdateAvailableToGet?.Invoke(this, updateStatus);
+	}
+
+	private async Task<bool> GetInstallerAsync(Version targetVersion)
+	{
+		if (CheckIfInstallerDownloaded())
+		{
+			return true;
+		}
+		using HttpRequestMessage message = new(HttpMethod.Get, ReleaseURL);
+		message.Headers.UserAgent.Add(new("WalletWasabi", "2.0"));
+		var response = await HttpClient.SendAsync(message).ConfigureAwait(false);
+
+		JObject jsonResponse = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+		string softwareVersion = jsonResponse["tag_name"]?.ToString() ?? throw new InvalidDataException("Endpoint gave back wrong json data or it's changed.");
+
+		// If the version we are looking for is not the one on github, somethings wrong.
+		Version githubVersion = new(softwareVersion[1..]);
+		Version shortGithubVersion = new(githubVersion.Major, githubVersion.Minor, githubVersion.Build);
+		if (targetVersion != shortGithubVersion)
+		{
+			throw new InvalidDataException("Target version from backend does not match with the latest github release. This should be impossible.");
+		}
+
+		// Get all asset names and download urls to find the correct one.
+		List<JToken> assetsInfos = jsonResponse["assets"]?.Children().ToList() ?? throw new InvalidDataException("Missing assets from response.");
+		List<string> assetDownloadUrls = new();
+		foreach (JToken asset in assetsInfos)
+		{
+			assetDownloadUrls.Add(asset["browser_download_url"]?.ToString() ?? throw new InvalidDataException("Missing download url from response."));
+		}
+
+		(string url, string fileName) = GetAssetToDownload(assetDownloadUrls);
+
+		var tmpFileName = Path.Combine(DownloadsDir, $"{fileName}.tmp");
+		var newFileName = Path.Combine(DownloadsDir, fileName);
+
+		// This should also be done using Tor.
+		using System.Net.Http.HttpClient httpClient = new();
+
+		// Get file stream and copy it to downloads folder to access.
+		using var stream = await httpClient.GetStreamAsync(url).ConfigureAwait(false);
+		Logger.LogInfo("Installer stream downloaded, copying...");
+		var file = File.OpenWrite(tmpFileName);
```

Comment:
All streams are disposable. Please go through the whole method and make sure u put `using` everywhere.